torch/csrc/autograd/python_variable.cpp

```diff
@@ -413,6 +427,189 @@ static PyObject * THPVariable_device(THPVariable* self) {
   END_HANDLE_TH_ERRORS
 }
 
+// Returns the type object of the most derived subclass of torch._C._TensorBase.
+PyTypeObject *THPVariable_result_ptype(PyObject *self, PyObject *args) {
+    HANDLE_TH_ERRORS
+    PyTypeObject *ptype;
+    PyObject *obj;
+    
+    //Initialize on First Call
+    bool initialized = false;
+    static PyObject *base, *parameter;
+    if (!initialized){
+        base = THPVariableClass;
+        parameter = PyObject_GetAttrString(PyImport_ImportModule("torch.nn.parameter"), "Parameter");
+        if (parameter == NULL)
+            return NULL;
+        initialized = true;
+    }
+
+    ptype = (PyTypeObject*)base;
+    Py_INCREF(ptype);
+    if (self != NULL && PyObject_IsInstance(self, base) && !PyObject_IsInstance(self, parameter) && PyObject_IsInstance(self, (PyObject*)ptype)){
+        ptype = self->ob_type;
+    }
+    if (args != NULL && PyTuple_Check(args)){
+        for (int i = 0; i < PyTuple_GET_SIZE(args); ++i){
+            obj = PyTuple_GET_ITEM(args, i);
+            if (PyObject_IsInstance(obj, base) && !PyObject_IsInstance(obj, parameter) && PyObject_IsInstance(obj, (PyObject*)ptype)){
+                ptype = obj->ob_type;
+            }
+        }
+    }
+    /* obj are borrowed references */
+    return ptype;
+    END_HANDLE_TH_ERRORS
+}
+
+// Implements the default __array_ufunc__ python subclass type propagation. Used in torch._C._TensorBase.
```

Comment:
We need some link to the specification of what `__array_ufunc__` does, or a recapitulation of its spec here.

Reply:
> "We need some link to the specification of what __array_ufunc__ does, or a recapitulation of its spec here."

` __array_ufunc__` traditionally does two things:
 1. Cast all input args and output kwargs to the np.NDarray base class
 2. Look for any subclass overrides of the ufunc that is being called. (eg. did the subclass override np.add?)
3. Cast all result args to the np.NDarray subclass

When `__array_func__` is implemented inside torch.Tensor we should note that torch.Tensor is NOT as subclass of np.NDArray. It should do the following:
 1. Cast all input and output args to the np.NDarray base class 
 2. DO NOT Look for any subclass overrides of the ufunc that is being called (It is NOT a subclass of np.NDArray and can not override ufuncs).
3. Cast all result args to the np.NDarray subclass

My implementation of `__array_ufunc__` does not support subclass overrides, of numpy ufuncs because Tensor is not a subclass of np.NDArray.

Reply:
OK, don't tell me here, put this in the code, if you can :)